pkg/providers/amifamily/bootstrap/nodeadm.go

```diff
@@ -0,0 +1,165 @@
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package bootstrap
+
+import (
+	"encoding/base64"
+	"fmt"
+	"strings"
+
+	admapi "github.com/awslabs/amazon-eks-ami/nodeadm/api"
+	admv1alpha1 "github.com/awslabs/amazon-eks-ami/nodeadm/api/v1alpha1"
+	"github.com/samber/lo"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/util/json"
+	"sigs.k8s.io/yaml"
+
+	"github.com/aws/karpenter-provider-aws/pkg/utils"
+)
+
+type Nodeadm struct {
+	Options
+}
+
+func (n Nodeadm) Script() (string, error) {
+	nodeConfigYAML, err := n.getNodeConfigYAML()
+	if err != nil {
+		return "", fmt.Errorf("generating NodeConfig, %w", err)
+	}
+	customEntries, err := n.parseUserData()
+	if err != nil {
+		return "", fmt.Errorf("parsing custom UserData, %w", err)
+	}
+	mimeArchive := utils.MIMEArchive(append([]utils.MIMEEntry{{
+		ContentType: utils.MIMEContentTypeNodeConfig,
+		Content:     nodeConfigYAML,
+	}}, customEntries...))
+	userData, err := mimeArchive.Serialize()
+	if err != nil {
+		return "", err
+	}
+	return userData, nil
+}
+
+func (n Nodeadm) getNodeConfigYAML() (string, error) {
+	config := &admv1alpha1.NodeConfig{
+		TypeMeta: metav1.TypeMeta{
+			Kind:       admapi.KindNodeConfig,
+			APIVersion: admv1alpha1.GroupVersion.String(),
+		},
+		Spec: admv1alpha1.NodeConfigSpec{
+			Cluster: admv1alpha1.ClusterDetails{
+				Name:              n.ClusterName,
+				APIServerEndpoint: n.ClusterEndpoint,
+				CIDR:              n.ClusterCIDR,
+			},
+		},
+	}
+	if lo.FromPtr(n.CABundle) != "" {
+		ca, err := base64.StdEncoding.DecodeString(*n.CABundle)
+		if err != nil {
+			return "", fmt.Errorf("decoding CABundle, %w", err)
+		}
+		config.Spec.Cluster.CertificateAuthority = ca
+	}
+	inlineConfig, err := n.generateInlineKubeletConfiguration()
+	if err != nil {
+		return "", err
+	}
+	config.Spec.Kubelet.Config = inlineConfig
+	config.Spec.Kubelet.Flags = n.getExtraKubeletFlags()
+
+	// Convert to YAML at the end for improved legibility.
+	configYAML, err := yaml.Marshal(config)
+	if err != nil {
+		return "", err
+	}
+	return string(configYAML), nil
+}
+
+// generateInlineKubeletConfiguration returns a serialized form of the KubeletConfiguration specified by the Nodeadm
+// options, for use with nodeadm's NodeConfig struct.
+func (n Nodeadm) generateInlineKubeletConfiguration() (map[string]runtime.RawExtension, error) {
+	kubeConfigJSON, err := json.Marshal(n.KubeletConfig)
+	if err != nil {
+		return nil, err
+	}
+	kubeConfigMap := map[string]interface{}{}
+	err = json.Unmarshal(kubeConfigJSON, &kubeConfigMap)
+	if err != nil {
+		return nil, err
+	}
+	kubeConfigMap = lo.Assign(kubeConfigMap, n.getExtraKubeletConfigurationFields())
+	inlineConfig := map[string]runtime.RawExtension{}
```

Comment:
I can leave some additional comments here, it might be possible to convert directly into a `map[string]runtime.RawExtension`, I honestly hadn't tried since I assumed it would not work. I'll give that a shot, but if it doesn't work I'll add some additional comments.

Reply:
Sounds good. I'm fine either way as long as we are sure of the direction here.